src/no-side-effects.ts

```diff
@@ -0,0 +1,125 @@
+// no-side-effects.ts
+
+/*
+ * Copyright (c) 2022-2024 Check Digit, LLC
+ *
+ * This code is licensed under the MIT license (see LICENSE.txt for details).
+ */
+
+import type { Node } from 'estree';
+import type { Rule } from 'eslint';
+
+interface RuleOptions {
+  excludedIdentifiers: string[];
+}
+
+function isAwaitExpression(statement: Node): boolean {
+  return statement.type === 'ExpressionStatement' && statement.expression.type === 'AwaitExpression';
+}
+
+function isCallExpressionCalleeMemberExpression(statement: Node, excludedIdentifiers: string[]): boolean {
+  return (
+    statement.type === 'ExpressionStatement' &&
+    statement.expression.type === 'CallExpression' &&
+    statement.expression.callee.type === 'MemberExpression' &&
+    statement.expression.callee.object.type === 'Identifier' &&
+    !excludedIdentifiers.includes(statement.expression.callee.object.name)
+  );
+}
+
+function isCallExpressionCalleeIdentifier(statement: Node, excludedIdentifiers: string[]): boolean {
+  return (
+    statement.type === 'ExpressionStatement' &&
+    statement.expression.type === 'CallExpression' &&
+    statement.expression.callee.type === 'Identifier' &&
+    !excludedIdentifiers.includes(statement.expression.callee.name)
+  );
+}
+
+function isVariableDeclarationAwaitExpression(statement: Node): boolean {
+  return (
+    statement.type === 'VariableDeclaration' &&
+    statement.declarations.length > 0 &&
+    statement.declarations[0]?.id !== undefined &&
+    statement.declarations[0].id.type === 'Identifier' &&
+    statement.declarations[0].init !== undefined &&
+    statement.declarations[0].init !== null &&
+    statement.declarations[0].init.type === 'AwaitExpression'
```

Comment:
i could be wrong, but it feels like the check for init can reuse the logic above. similarly for the following 2 variable declaration related init checking as well.

Reply:
Somehow it likes eslint traversal only if we have the above checks and can't be reused.